controllers/users.ts

```diff
@@ -0,0 +1,82 @@
+import { NextFunction, Request, Response } from 'express';
+import httpStatus from 'http-status';
+import { isValidObjectId } from 'mongoose';
+import { userModel } from '../models/users';
+
+export const createUser = async (request: Request, response: Response, next: NextFunction) => {
+	const data = request.body;
+
+	try {
+		const newUser = await userModel.create(data);
+		response.status(httpStatus.CREATED).send({ ...newUser, password: undefined });
+	} catch (error) {
+		next(error);
+	}
+};
+
+export const getUsers = async (
+	request: Request<{}, {}, {}, { userId?: string; username?: string; email?: string; password: never }>,
+	response: Response,
+	next: NextFunction
+) => {
+	const { userId } = request.query;
+
+	if (!!userId && !isValidObjectId(userId)) {
+		response.status(httpStatus.BAD_REQUEST).send(`Invalid id ${userId}`);
+		return;
+	}
+
+	const filters = Object.entries(request.query)
+		.filter(([key, value]) => Object.keys(userModel.schema.obj).includes(key) && value !== undefined)
+		.reduce((previous, [key, value]) => ({ ...previous, [key]: value }), {});
+
+	if (Object.entries(filters).length === 0 && userId === undefined) {
+		response.status(httpStatus.BAD_REQUEST).send('No valid filters provided (username, email, userId)');
+		return;
+	}
+
+	try {
+		const user = await userModel.findOne(filters).select('-password');
+		response.status(httpStatus.OK).send(user);
+	} catch (error) {
+		next(error);
+	}
+};
+
+export const updateUserById = async (request: Request<{ id: string }>, response: Response, next: NextFunction) => {
+	const { id: userId } = request.params;
+	const data = request.body;
```

Comment:
check `Object.keys(data || {}) > 0`, apparently if u don't provide data it doesn't throw an error for non-existing ID